### PR TITLE
Pythia EvtGen interface 

### PR DESCRIPTION
**Pull request by mchrzasz**

### CODE REVIEW

**Review by tmadlener — CHANGES_REQUESTED** [review]
Hi @mchrzasz

Thanks for opening this. I haven't yet tested it, but I had a look at the code and I have added a few comments requesting some small changes.

From the looks of it the `PythiaEvtGen_Interface` class does not depend on anything from key4hep, so this can probably be moved quite easily in the end, right? Or did I miss something here?

**Comment by vvolkl** [comment]
Thanks @mchrzasz for the PR and @tmadlener for the comments. I mostly think that there are still some tests needed, but I propose to merge this now and adress the comments in a future PR.

**Comment by clementhelsens** [comment]
@vvolkl we should not merge now, @mchrzasz is fixing an issue that I discovered when trying to decay Bc hadrons. in this case the generation needs to be done until a ccbar pair is produced at ME level

**Comment by mchrzasz** [comment]
Hey,

Yes I think I will fix the ccbar issue today with all other comments.

File: standalone/src/PythiaEvtGen_Interface.cc [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+
+  
+}
+//=============================================================================
+// Destructor
+//=============================================================================
+PythiaEvtGen_Interface::~PythiaEvtGen_Interface()
+{
+  delete pythia;
```
[the remark this answers is not in the document]
Comment (mchrzasz) [reply standalone/src/PythiaEvtGen_Interface.cc:59]: Yes you are right. The ownership is not transferred. I have deleted this deletion :)

File: standalone/src/DelphesPythia8EvtGenReader_k4Interface.h [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+#include <iostream>
+#include <memory>
+
+
+#include "PythiaEvtGen_Interface.h"
+
+//---------------------------------------------------------------------------
+
+
+class DelphesPythia8EvtGenReader: public DelphesInputReader {
```
[the remark this answers is not in the document]
Comment (mchrzasz) [reply standalone/src/DelphesPythia8EvtGenReader_k4Interface.h:36]: Yes, you are totally right :) Changed to DelphesPythia8EvtGenReader_k4Interface

File: standalone/src/PythiaEvtGen_Interface.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  //  std::vector<EvtGen*> EG_gens;
+  EvtGen* evtgen;
+  std::vector<int> motherIDs;
+  std::vector<std::string> sig_names;
+  Pythia8::Pythia *pythia;
+  std::string particleDataFile;
+  int init_seed;
+ 
+  std::string pdl;
+  int **signal_map;
```
[the remark this answers is not in the document]
Comment (mchrzasz) [reply standalone/src/PythiaEvtGen_Interface.h:68]: Changed to vector of vectors.

File: standalone/src/PythiaEvtGen_Interface.cc [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+///////////////////////////////////////////////////////////////////////////////
+
+
+bool PythiaEvtGen_Interface::check_Signal_Appereance()
+{
+  unsigned NOfSignal=motherIDs.size();
+  NOfSignal_list = std::vector(NOfSignal, 0);
+  //signal_map= new int[NOfSignal][10];// here we assume we won't have more then 10 signal candidates
+
+  signal_map= new int*[NOfSignal];
```
[the remark this answers is not in the document]
Comment (mchrzasz) [reply standalone/src/PythiaEvtGen_Interface.cc:123]: This is obsolete now with the vector of vectors.

File: standalone/src/PythiaEvtGen_Interface.cc [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+  
+}
+//=============================================================================
+// Destructor
+//=============================================================================
+PythiaEvtGen_Interface::~PythiaEvtGen_Interface()
+{
+  delete pythia;
+  delete evtgen;
+  delete signal_map;
```
[the remark this answers is not in the document]
Comment (mchrzasz) [reply standalone/src/PythiaEvtGen_Interface.cc:61]: Correct. Now it's gone

**Review by tmadlener — COMMENTED** [review]
Hi @mchrzasz thanks for the fixes and updates. I have two additional comments, that are a bit nit-picky and do not necessarily have to go into this version, but can also be fixed later.

If I understand correctly now also the ccbar issue that @clementhelsens mentioned, right? For me this is good to be merged as it is, but I will let him comment, on whether everything works as expected right now.

File: standalone/src/PythiaEvtGen_Interface.cc [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+  EvtRandomEngine* eng = 0;
+  //eng = new EvtMTRandomEngine(seed);  // look out, this works only with C++11, OLD C++ use: eng = new EvtSimpleRandomEngine()
+  eng=new EvtSimpleRandomEngine();
+  EvtRandom::setRandomEngine( eng );
+
+  if(debug) std::cout<<"Creating EvtGen instance"<<std::endl;
+
+  // Create the EvtGen generator object, passing the external generators
+
+  evtgen = new EvtGen(decayfile.c_str(), pdl.c_str(), eng, radCorrEngine, &extraModels);
```
[the remark this answers is not in the document]
Comment (mchrzasz) [reply standalone/src/PythiaEvtGen_Interface.cc:105]: I think (to be checked) that EvtGen doesn't take the ownership. So indeed I can change to unique pointer.
About the ccbar: not yet. Next commit will fix that :) Will let you know

[the remark this answers is not in the document]
Comment (clementhelsens) [reply standalone/src/PythiaEvtGen_Interface.cc:105]: @tmadlener , Marcin, just ping me when the changes for c flavour hadrons is there and I can  give it a try

**Comment by vvolkl** [comment]
Thanks everyone, I'm going to merge this now as we want to do a build that includes all edm4hep covariance matrix fixes.

**Comment by mchrzasz** [comment]
I will now create a 2nd PR. I have just finished testing the ccbar solution. it's very simple fix so I would merge it as well.

**Pull request merged**